django_plastic_tickets/models.py

```diff
@@ -42,42 +58,90 @@ class ProductionMethod(DescribedModel):
         return f'{self.name} ({self.production_method})'
 
 
-class MaterialColor(DescribedModel):
-    pass
+class Manufacturer(DescribedModel):
+    url = models.URLField(blank=True, null=True)
 
 
 class Material(models.Model):
-    """A physical material that is/was in Stock"""
-    color = models.ForeignKey(MaterialColor, on_delete=models.CASCADE)
+    """Base class for a physical material"""
     type = models.ForeignKey(MaterialType, on_delete=models.CASCADE)
+    manufacturer = models.ForeignKey(Manufacturer, on_delete=models.CASCADE)
     name = models.TextField()
-    url = models.URLField()
-    optimal_temp = models.FloatField()
-    min_temp = models.FloatField()
-    max_temp = models.FloatField()
+    url = models.URLField(blank=True, null=True)
+    ral_color_number = models.PositiveIntegerField(
+        validators=[validate_ral_number],
+        help_text=gettext('A RAL classic color number'))
+
+    def to_dict(self) -> dict:
+        return {'display_name': (self.get_color_name() +
+                                 f' (RAL {self.ral_color_number})'),
+                'ral_number': self.ral_color_number}
+
+    def get_color_name(self):
+        lang = translation.get_language()
+        return ral_colors[self.ral_color_number][lang]
+
+    def __str__(self) -> str:
+        return f'{self.name} ({self.type.name}, {self.get_color_name()})'
+
+
+class MaterialResin(Material):
+    pass
+
+
+class MaterialFilament(Material):
+    nozzle_temp = models.FloatField(blank=True, null=True)
+    bed_temp = models.FloatField(blank=True, null=True)
+    diameter = models.FloatField(blank=True, null=True, default=1.75)
+
+
+class MaterialPackage(models.Model):
+    """A spool, bottle or any other material package"""
+    manufacturer = models.ForeignKey(Manufacturer, on_delete=models.CASCADE)
+    empty_weight = models.FloatField(blank=True, null=True)
+    usual_capacity = models.FloatField(blank=True, null=True)
+    appearance = models.TextField()
 
     def __str__(self) -> str:
-        return f'{self.name} ({self.type.name}, {self.color.name})'
+        return (f'{self.manufacturer}, ' +
+                f'{self.usual_capacity} g,i ' +
+                f'{self.appearance}')
 
 
 class MaterialStock(models.Model):
+    """Base class for a material that is/was in stock"""
     material = models.ForeignKey(Material, on_delete=models.CASCADE)
+    package = models.ForeignKey(MaterialPackage, on_delete=models.CASCADE)
     label = models.PositiveIntegerField(
         help_text=gettext('The internal label to identify the material'),
         unique=True)
-    consumed = models.BooleanField(default=False)
+    initial_material_weight = models.FloatField(help_text=gettext(
+        'Initial weight of the material in gram, ' +
+        'not including the weight of the package'))
+    current_weight = models.FloatField(help_text=gettext(
+        'Current weight of the material in gram, ' +
+        'including the weight of the package'))
```

Comment:
This might be confusing at the beginning, but you know the initial weight of the material on a filament roll or resin bottle, but you can only measure the weight of it with the package.

Reply:
I am agreeing that `current_weight` should contain the package weight but then you also have to store the **weight** of the **package** in some way, otherwise the `initial_material_weight` would solve no purpose.

I guess this was your intention with `MaterialPackage`?     
So would some inventory workflow be this:
- Add a `MaterialPackage` for a light, transparent filament spool (from Extrudr)
- Add a `Material` for brown GreenTEC from Extrudr
- For every new filament in stock create a `MaterialStock` with the weight written on the package as `initial_material_weight` and the `current_weight = empty_weight + initial_material_weight`
- Calculate the remaining weight at any time with `current_weight - empty_weight`
Can you ACK or update this example?

One side effect of this is, that we are assuming an equal weight of every package.

Another small thing I have seen is that both `Material` and `MaterialPackage` are storing a `Manufacturer`. Is this intentional?

Reply:
- Yes, this is what the typical workflow would look like.
- I think it is safe to assume that every package which looks the same has also the same weight. One can still weight a new package with material and subtract the weight of the material (which is known from the manufacturer) to check that.
- It is intentional that both `Material` and `MaterialPackage` store `Manufacturer`. This helps to distinguish the packages.